### PR TITLE
fix: nil pointer dereference in HealthStrategy

### LINKED ISSUE #801: [Bug]: HealthStrategy tries to dereference a null Health pointer
### Testcontainers version

0.17.0

### Using the latest Testcontainers version?

Yes

### Host OS

Ubuntu 22.04

### Host arch

amd64

### Go version

1.19

### Docker version

```shell
Client: Docker Engine - Community
 Version:           23.0.0
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        e92dd87
 Built:             Wed Feb  1 17:47:51 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.0
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       d7573ab
  Built:            Wed Feb  1 17:47:51 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.16
  GitCommit:        31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.15.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 5
  Running: 2
  Paused: 0
  Stopped: 3
 Images: 578
 Server Version: 23.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-58-generic
 Operating System: Ubuntu 22.04.1 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.29GiB
 Name: gondor
 ID: DVVB:WRMF:MXNZ:2HZ3:JYZC:BQSD:6E7X:RY7E:IHRI:SV3A:ERUG:KTPF
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: massenz
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  192.168.99.101:30500
  127.0.0.0/8
 Live Restore Enabled: false
```

### What happened?

I used the code exactly as in the docs (and other variations too) and the tests failed with a panic (see log):

```
	req := testcontainers.ContainerRequest{
		Image:        OpaImage,
		ExposedPorts: []string{OpaPort},
		WaitingFor: wait.ForHealthCheck().
			WithPollInterval(1 * time.Second),
	}
```

### Relevant log output

```shell
└─( ko ./gentests/internals

Running Suite: Internals Suite
==============================
Random Seed: 1675555046
Will run 1 of 1 specs

2023/02/04 15:57:27 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 23.0.0
  API Version: 1.42
  Operating System: Ubuntu 22.04.1 LTS
  Total Memory: 32045 MB
2023/02/04 15:57:27 Starting container id: 286d87d82a44 image: docker.io/testcontainers/ryuk:0.3.4
2023/02/04 15:57:27 Waiting for container id 286d87d82a44 image: docker.io/testcontainers/ryuk:0.3.4
2023/02/04 15:57:28 Container is ready id: 286d87d82a44 image: docker.io/testcontainers/ryuk:0.3.4
2023/02/04 15:57:28 Starting container id: f6de79a5282d image: openpolicyagent/opa:0.48.0
2023/02/04 15:57:28 Waiting for container id f6de79a5282d image: openpolicyagent/opa:0.48.0
•! Panic [1.011 seconds]
Gentests/Internals/Server
/home/marco/Development/CopilotIQ/opa-tests/gentests/internals/server_test.go:10
  when the tests start
  /home/marco/Development/CopilotIQ/opa-tests/gentests/internals/server_test.go:12
    should have a running OPA server [It]
    /home/marco/Development/CopilotIQ/opa-tests/gentests/internals/server_test.go:13

    Test Panicked
    runtime error: invalid memory address or nil pointer dereference
    /opt/go/src/runtime/panic.go:260

    Full Stack Trace
    github.com/testcontainers/testcontainers-go/wait.(*HealthStrategy).WaitUntilReady(0xc000120000, {0xb72898?, 0xc00002e110?}, {0xb73d60, 0xc000000320})
    	/home/marco/.local/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.17.0/wait/health.go:79 +0xd9
    github.com/testcontainers/testcontainers-go.(*DockerContainer).Start(0xc000000320, {0xb72898, 0xc00002e110})
    	/home/marco/.local/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.17.0/docker.go:198 +0x24f
    github.com/testcontainers/testcontainers-go.GenericContainer({_, _}, {{{{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, 0x0, 0x0, ...}, ...}, ...})
    	/home/marco/.local/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.17.0/generic.go:75 +0x3d4
    github.com/CopilotIQ/opa-tests/gentests/internals.NewOpaContainer({0xb72898, 0xc00002e110}, {0xa87143, 0xc})
    	/home/marco/Development/CopilotIQ/opa-tests/gentests/internals/server.go:36 +0x1ae
    github.com/CopilotIQ/opa-tests/gentests/internals_test.glob..func1.1.1()
    	/home/marco/Development/CopilotIQ/opa-tests/gentests/internals/server_test.go:14 +0x45
    github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc00003c4e0?)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/leafnodes/runner.go:113 +0xb1
    github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0x417610?)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/leafnodes/runner.go:64 +0x125
    github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0x400?)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/leafnodes/it_node.go:26 +0x7b
    github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc0003ac000, 0xc000071af0?, {0xb6d600, 0xc000300ec0})
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/spec/spec.go:215 +0x2a9
    github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc0003ac000, {0xb6d600, 0xc000300ec0})
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/spec/spec.go:138 +0xf2
    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc000360780, 0xc0003ac000)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/specrunner/spec_runner.go:200 +0xf1
    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc000360780)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/specrunner/spec_runner.go:170 +0x1b6
    github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc000360780)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/specrunner/spec_runner.go:66 +0xc5
    github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc0002f6690, {0x7f263aa33ca0, 0xc0000d6d00}, {0xa896b1, 0xf}, {0xc00035e8f0, 0x1, 0x1}, {0xb73360, 0xc000300ec0}, ...)
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/internal/suite/suite.go:62 +0x4c5
    github.com/onsi/ginkgo.RunSpecsWithCustomReporters({0xb6e2a0?, 0xc0000d6d00}, {0xa896b1, 0xf}, {0xc000060718, 0x1, 0xc000060728?})
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/ginkgo_dsl.go:226 +0x189
    github.com/onsi/ginkgo.RunSpecs({0xb6e2a0, 0xc0000d6d00}, {0xa896b1, 0xf})
    	/home/marco/.local/go/pkg/mod/github.com/onsi/ginkgo@v1.12.1/ginkgo_dsl.go:207 +0x10b
    github.com/CopilotIQ/opa-tests/gentests/internals_test.TestInternals(0x0?)
    	/home/marco/Development/CopilotIQ/opa-tests/gentests/internals/internals_suite_test.go:12 +0x85
    testing.tRunner(0xc0000d6d00, 0xac79e0)
    	/opt/go/src/testing/testing.go:1446 +0x10b
    created by testing.(*T).Run
    	/opt/go/src/testing/testing.go:1493 +0x35f
------------------------------


Summarizing 1 Failure:

[Panic!] Gentests/Internals/Server when the tests start [It] should have a running OPA server 
/opt/go/src/runtime/panic.go:260

Ran 1 of 1 Specs in 1.011 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestInternals (1.01s)
FAIL

Ginkgo ran 1 suite in 2.027766693s
Test Suite Failed
```

### Additional information

This is caused by [this line](https://github.com/testcontainers/testcontainers-go/blob/a201e1d72d529a4d252e38702fc1718ba761cd16/wait/health.go#L79):

```
if state.Health.Status != "healthy" {
```
due to `Health` being `nil` (at least, initially) - however, the following should work (I've rewritten the strategy and tested to work, but it's been a very limited testing on just that one container):

```
if state.Health == nil || state.Health.Status != "healthy" {
```

Whether this is "intended" I am not sure (but it achieves what I want, to wait for the container to be ready).

In any event, as a matter of best practice, before dereferencing "two levels deep" it would be better to defend against `nil` values.

### PR DESCRIPTION
**Pull request by massenz**
## What does this PR do?

Fixes a panic caused by a `nil` pointer dereference in the `HealthStrategy.WaitUntilReady` function.
See Issue #801 

## Why is it important?
Anyone using the `HealthStrategy` for their tests would see them failing with the panic.

## Related issues

- Closes #801

## How to test this PR
I have added the following tests, to cover the various scenarios:
```
PASS wait.TestWaitForHealthTimesOutForUnhealthy (0.20s)
PASS wait.TestWaitForHealthSucceeds (0.00s)
PASS wait.TestWaitForHealthWithNil (0.30s)
```
The `TestWaitForHealthWithNil` test will cause the panic with the original code, but it does not after the PR is applied.

### CODE REVIEW

**Review by mdelapenya — COMMENTED** [review]
I'm not sure about this 🤔 Do we want to eventually pass the test if the health is nil? Which comes with a second question: what do we want to do in the health strategy when the health is nil? Fail/Continue/Log?

EDIT: the above line **was** for this block: https://github.com/testcontainers/testcontainers-go/pull/802/files#diff-62f25e9df2042653e9484951d6537c4088f6580206cbaef6bb30d610bfdd6befR69-R74. So please do not get me wrong 🙏  I think we should check for nil, although considering different states.

In the current implementation we are considering that not having declared a Health in the container is exactly the same as not being healthy (we sleep). I'm not opposed to that, but what to discuss with you about it: should be consider not having Health in the state as healthy or not?

**Comment by massenz** [comment]
> I'm not sure about this thinking Do we want to eventually pass the test if the health is nil? Which comes with a second question: what do we want to do in the health strategy when the health is nil? Fail/Continue/Log?

Hey, thanks for super-quick review and sorry for not explaining the tests better (I've now added comments, but please let me know if still not sufficient).
And, no, absolutely, the thinking is that if `Health` is `nil` the strategy will (eventually, after the timeout) fail

 
> EDIT: the above line **was** for this block: https://github.com/testcontainers/testcontainers-go/pull/802/files#diff-62f25e9df2042653e9484951d6537c4088f6580206cbaef6bb30d610bfdd6befR69-R74. So please do not get me wrong pray I think we should check for nil, although considering different states.

Indeed - in fact that test is meant to "emulate" observed container behavior: initially `Health` is `nil` and then it becomes healthy: in the original code, this would have caused a panic, but now it waits and (if the state becomes `healthy` then it succeeds).

I've added another test (`TestWaitFailsForNilHealth`) where Health stays `nil` and the strategy eventually times out and returns the (expected) `context.DeadlineExceeded` error.

 
> In the current implementation we are considering that not having declared a Health in the container is exactly the same as not being healthy (we sleep). I'm not opposed to that, but what to discuss with you about it: should be consider not having Health in the state as healthy or not?

Yes, this follows the same principle, as mentioned above, but it waits for it to either become non-nil, and if not it fails - if it does become non-nil, then it checks the state.

Hopefully that clarifies, and the additional test proves that.

**Review by mdelapenya — COMMENTED** [review]
Just a tiny suggestion in the new test method. Once we discuss on that, LGTM

File: wait/health_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	wg := NewHealthStrategy().
+		WithStartupTimeout(500 * time.Millisecond).
+		WithPollInterval(100 * time.Millisecond)
+
+	err := wg.WaitUntilReady(context.Background(), target)
+	if err != nil && err == context.DeadlineExceeded {
+		return
+	}
+	// Any error, or success will cause this test to fail
+	t.Fatal(err)
```
[the remark this answers is not in the document]
Comment (massenz) [reply wait/health_test.go:99]: :+1:

File: wait/health_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+// time out.
+func TestWaitForHealthTimesOutForUnhealthy(t *testing.T) {
+	target := healthStrategyTarget{Health: &types.Health{Status: types.Unhealthy}}
+	wg := NewHealthStrategy().WithStartupTimeout(100 * time.Millisecond)
+	err := wg.WaitUntilReady(context.Background(), target)
+
+	if err != nil && err == context.DeadlineExceeded {
+		return
+	}
+	t.Fatal(err)
```
Comment (mdelapenya) [inline wait/health_test.go:54]: Let's rewrite this block as below:
```suggestion
	assert.NotNil(t, err)
	assert.True(t, errors.Is(err, context.DeadlineExceeded))
```

File: wait/health_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+// TestWaitForHealthSucceeds ensures that a healthy container always succeeds.
+func TestWaitForHealthSucceeds(t *testing.T) {
+	target := healthStrategyTarget{Health: &types.Health{Status: types.Healthy}}
+	wg := NewHealthStrategy().WithStartupTimeout(100 * time.Millisecond)
+	err := wg.WaitUntilReady(context.Background(), target)
+
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (mdelapenya) [inline wait/health_test.go:65]:
```suggestion
	assert.Nil(t, err)
```

File: wait/health_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		// wait a bit to simulate startup time and give check time to at least
+		// try a few times with a nil Health
+		time.Sleep(200 * time.Millisecond)
+		target.Health = &types.Health{Status: types.Healthy}
+	}(target)
+
+	err := wg.WaitUntilReady(context.Background(), target)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (mdelapenya) [inline wait/health_test.go:86]:
```suggestion
	assert.Nil(t, err)
```

**Review by mdelapenya — APPROVED** [review]
I'm approving the PR even without the suggestions I added on the style on how to assert.

Will merge it once you approve them or discard them for a different preference in the assert style

Thanks for t¡your time and patience contributing to the project 🚀

**Comment by massenz** [comment]
> Just a tiny suggestion in the new test method. Once we discuss on that, LGTM

oh, I'm so sorry!
I thought you meant only that "new test method" and didn't realize I should make the changes everywhere
(I usually use Gingko/Gomega assertions, not really used to the "vanilla" Go testing)

On an unrelated note - I have the impression that some containers (most? all?) do not really set the `Health` value: I can see the `running` is `true`, but `Health` stays nil.
The code is still valid, but I'm now wondering about the value of this particular Strategy.

BTW - while working on this, I was wondering adding an `HealthEndpointStrategy` which checks on one of the `mapppedPorts` for a particular endpoint (default: `/health`) makes a GET request and expects a given status code (default `http.StatusOk`)
What do you think?

**Comment by mdelapenya** [comment]
> > Just a tiny suggestion in the new test method. Once we discuss on that, LGTM
> 
> oh, I'm so sorry! I thought you meant only that "new test method" and didn't realize I should make the changes everywhere (I usually use Gingko/Gomega assertions, not really used to the "vanilla" Go testing)

No worries, I totally understand :) I was not explicit, so not an issue at all

> On an unrelated note - I have the impression that some containers (most? all?) do not really set the `Health` value: I can see the `running` is `true`, but `Health` stays nil. The code is still valid, but I'm now wondering about the value of this particular Strategy.

The State function, when it's called from a container, inspects the container, so it will take whatever it's in the Dockerfile. If the container you use does not have a Health it's because it's not there.

```golang
// State returns container's running state
func (c *DockerContainer) State(ctx context.Context) (*types.ContainerState, error) {
	inspect, err := c.inspectRawContainer(ctx)
	if err != nil {
		if c.raw != nil {
			return c.raw.State, err
		}
		return nil, err
	}
	return inspect.State, nil
}
```

> BTW - while working on this, I was wondering adding an `HealthEndpointStrategy` which checks on one of the `mapppedPorts` for a particular endpoint (default: `/health`) makes a GET request and expects a given status code (default `http.StatusOk`) What do you think?

For this, you can use wait.FoHTTP https://golang.testcontainers.org/features/wait/http/ which, I think, comes with what you expect: check for status code, check in an HTTP endpoint, and check for a response. Is that what you are looking for?

File: wait/health_test.go (hunk 3 above)
Comment (mdelapenya) [reply wait/health_test.go:63]: Should we include this too?

**Comment by massenz** [comment]
> The State function, when it's called from a container, inspects the container, so it will take whatever it's in the Dockerfile. If the container you use does not have a Health it's because it's not there.

Thanks, Manuel - I didn't know of the `HEALTHCHECK` directive, thanks for making me discover it (I thought the health was somehow defined by the docker daemon according to some internal state).
As a matter of fact, OPA (Open Policy Agent) doesn't set it.

> For this, you can use wait.ForHTTP https://golang.testcontainers.org/features/wait/http/ which, I think, comes with what you expect: check for status code, check in an HTTP endpoint, and check for a response. Is that what you are looking for?

Well, honestly I have no idea how I missed that one :man_facepalming: it's not even like there's a want of options!!! :+1: 
Thanks!

Have committed your suggestion too.

**Review by mdelapenya — APPROVED** [review]
Great job during the review! I very much appreciate your effort in contributing the fix

image
